multi-layer-perceptron/jax_test.py

```diff
@@ -0,0 +1,308 @@
+
+# Suppress warning, info and error messages from jax, tf
+os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
+
+# Specify directory for data 
+DATA_DIR = './datasets'
+
+class MLPModel:
+
+    def __init__(self, layer_sizes, num_epochs=1, batch_size=128, n_targets=10, learning_rate=0.01):
+
+        # Array of each layer size
+        self.layer_sizes = layer_sizes
+
+        # Also refered to as the step size
+        # Scaling apply to estimated weight error during backpropagation
+        self.learning_rate = learning_rate
+
+        # Configure parameters
+        self.num_epochs = num_epochs
+        self.batch_size = batch_size
+        self.n_targets = n_targets
+
+        # Initialize all layers with random values
+        self.params = self.__init_layers()
+
+    def __init_layer_params(self, m, n, key, scale=1e-2):
+        """
+        A helper function to randomly initialize weights and biases
+        for a dense neural network layer.
+        """
+        w_key, b_key = random.split(key)
+        
+        return scale * random.normal(w_key, (n, m)), scale * random.normal(b_key, (n,))
+
+    def __init_layers(self, key=random.PRNGKey(0) ):
+        """
+        Initialize all layers for a fully-connected neural network.
+        """
+        sizes = self.layer_sizes
+        keys = random.split(key, len(sizes))
+
+        return [self.__init_layer_params(m, n, k) for m, n, k in zip(sizes[:-1], sizes[1:], keys)]
+
+def relu(x):
+    """
+    Rectified Linear Operation.
+    """
+    return jnp.maximum(0, x)
+
+def one_hot_encoder(x, k, dtype=jnp.float32):
+    """
+    Create a one-hot encoding of x of size k.
+    """
+    return jnp.array(x[:, None] == jnp.arange(k), dtype)
+
+def predict(params, input):
+    """
+    Run a single prediction on the input. 
+
+    Parameters: 
+        input (array): A n-dimensional array, typically a 1D or 2D array. For images for instance this is a 2D array. 
+    """
+
+    activations = input
+    for w, b in params[:-1]:
+        outputs = jnp.dot(w, activations) + b
+        activations = relu(outputs)
+    
+    final_w, final_b = params[-1]
+    logits = jnp.dot(final_w, activations) + final_b
+    
+    # Compute log softmax loss
+    loss = logits - logsumexp(logits)
+
+    return loss
+
+# Temp, move this
+batched_predict = vmap(predict, in_axes=(None, 0))
+
+def loss(params, inputs, targets):
+    """
+    Computes negative log likehood loss. 
+
+    Each prediction of batched_predict retuns the log softmax loss.
+    Computing the mean and returning the mean is this form of loss. 
+    """
+    preds = batched_predict(params, inputs)
+    loss = -jnp.mean(preds * targets)
+
+    return loss
+
+@jit
+def update(params, x, y, learning_rate):
+    """
+    Perform backpropagation using stochastic gradient descent.
+    """
+    grads = grad(loss)(params, x, y)
+
+    return [(w - learning_rate * dw, b - learning_rate * db) for (w, b), (dw, db) in zip(params, grads)]
```

Comment:
SGD still, what did you decide to do about Adam? Write your own or use the implementation from Flax or Haiku?

Reply:
Implementat adam using optax - essential another packages built of off jax for optimizers. 

RIght now the script is configured so that it can easily run either gradient descent or adam, with adam being the default its setup for. 

I think going forward I will look into implementing adam in jax from scratch to be more inline with the ground up implementation and style of Jax, but for now I am gonna have it like this so its wrapped up.